File: src/main/java/voidchess/ui/Table.java
package voidchess.ui;

import voidchess.board.ChessGameInterface;
import voidchess.board.MoveResult;
import voidchess.helper.*;
import voidchess.player.PlayerInterface;

/**
 * @author stephan
 */
public class Table implements ChessGameSupervisor, TableInterface {
    private ChessGameInterface game;
    private ChessboardComponent ui;
    private Chess960Panel panel960;
    private PlayerInterface whitePlayer, blackPlayer;
    private boolean whitePlayersTurn, resign;
    private ChessPanel parent;


    Table(ChessGameInterface game, ChessboardComponent ui, ChessPanel parent, Chess960Panel panel960) {
        this.game = game;
        this.ui = ui;
        this.parent = parent;
        this.panel960 = panel960;
    }

    public PawnPromotion askForPawnChange(Position pawnPosition) {
        if (whitePlayersTurn) return whitePlayer.askForPawnPromotionType(pawnPosition);
        else return blackPlayer.askForPawnPromotionType(pawnPosition);
    }

    @Override
    public void move(Move move) {
        MoveResult endoption;
        synchronized (this) {
            if (resign) {
                stopGame(MoveResult.RESIGN);
                return;
            }
            endoption = game.move(move);
            ExtendedMove extendedMove = game.getLastExtendedMove();
            ui.repaintAfterMove(extendedMove);

            whitePlayersTurn = !whitePlayersTurn;
        }
        if (endoption == MoveResult.NO_END) {
            boolean gaveCheck = game.isCheck(whitePlayersTurn);
            if (whitePlayersTurn) {
                whitePlayer.play();
                if(gaveCheck) {
                    blackPlayer.gaveCheck();
                }
            } else {
                blackPlayer.play();
                if(gaveCheck) {
                    whitePlayer.gaveCheck();
                }
            }
        } else {
            stopGame(endoption);
        }
    }

    @Override
    public void setWhitePlayer(PlayerInterface player) {
        whitePlayer = player;
        player.setColor(true);
    }

    @Override
    public void setBlackPlayer(PlayerInterface player) {
        blackPlayer = player;
        player.setColor(false);
    }

    @Override
    public void switchPlayer() {
        PlayerInterface formerWhitePlayer = whitePlayer;
        setWhitePlayer(blackPlayer);
        setBlackPlayer(formerWhitePlayer);
    }

    @Override
    public void startGame() {
        whitePlayersTurn = true;
        resign = false;
        game.initGame(panel960.getPositionCode());
        ui.repaintAtOnce();

        whitePlayer.gameStarts();
        blackPlayer.gameStarts();

        whitePlayer.play();
    }

    @Override
    public void stopGame(MoveResult endoption) {
        whitePlayer.gameEnds(endoption, !whitePlayersTurn);
        blackPlayer.gameEnds(endoption, !whitePlayersTurn);
        parent.gameover();
    }
}

File: src/main/kotlin/voidchess/ui/Table.kt
package voidchess.ui

import voidchess.board.ChessGameInterface
import voidchess.board.MoveResult
import voidchess.helper.*
import voidchess.player.PlayerInterface


class Table constructor(
        private val game: ChessGameInterface,
        private val ui: ChessboardComponent,
        private val parent: ChessPanel,
        private val panel960: Chess960Panel
) : ChessGameSupervisor, TableInterface {

    private lateinit var whitePlayer: PlayerInterface
    private lateinit var blackPlayer: PlayerInterface
    private var whitePlayersTurn: Boolean = false
    private var resign: Boolean = false

    override fun askForPawnChange(pawnPosition: Position): PawnPromotion {
        return if (whitePlayersTurn)
            whitePlayer.askForPawnPromotionType(pawnPosition)
        else
            blackPlayer.askForPawnPromotionType(pawnPosition)
    }

    override fun move(move: Move) {
        val endOption = synchronized(this) {
            if (resign) {
                stopGame(MoveResult.RESIGN)
                return
            }
            val endOption = game.move(move)
            val extendedMove = game.getLastExtendedMove()
            ui.repaintAfterMove(extendedMove)

            whitePlayersTurn = !whitePlayersTurn
            endOption
        }

        if (endOption === MoveResult.NO_END) {
            val gaveCheck = game.isCheck(whitePlayersTurn)
            if (whitePlayersTurn) {
                whitePlayer.play()
                if (gaveCheck) {
                    blackPlayer.gaveCheck()
                }
            } else {
                blackPlayer.play()
                if (gaveCheck) {
                    whitePlayer.gaveCheck()
                }
            }
        } else {
            stopGame(endOption)
        }
    }

    override fun setWhitePlayer(player: PlayerInterface) {
        whitePlayer = player
        player.setColor(true)
    }

    override fun setBlackPlayer(player: PlayerInterface) {
        blackPlayer = player
        player.setColor(false)
    }

    override fun switchPlayer() {
        val formerWhitePlayer = whitePlayer
        setWhitePlayer(blackPlayer)
        setBlackPlayer(formerWhitePlayer)
    }

    override fun startGame() {
        whitePlayersTurn = true
        resign = false
        game.initGame(panel960.positionCode)
        ui.repaintAtOnce()

        whitePlayer.gameStarts()
        blackPlayer.gameStarts()

        whitePlayer.play()
    }

    override fun stopGame(endoption: MoveResult) {
        whitePlayer.gameEnds(endoption, !whitePlayersTurn)
        blackPlayer.gameEnds(endoption, !whitePlayersTurn)
        parent.gameover()
    }
}
